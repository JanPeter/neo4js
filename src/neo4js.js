import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
