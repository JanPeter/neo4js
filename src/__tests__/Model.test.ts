// @flow

import neo4js, { Model, ModelInstance } from "../index";
import {
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from "../index";

type Props = {
  name: StringProperty,
  age?: NumberProperty,
};

class PersonInstance extends ModelInstance<Props> {}
class PersonModel extends Model<Props, PersonInstance> {}
const Person = new PersonModel("Person");

describe("Model", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  describe("create", () => {
    it("should create a instance of specific model", async () => {
      const paul: PersonInstance = await Person.create({
        name: "Paul",
        age: 21,
      });
      delete paul.props.guid;
      expect(paul).toMatchSnapshot();
    });
  });

  describe("findByGuid", () => {
    it("should get a instance with specific guid", async () => {
      let p1 = await Person.create({ name: "Paul" });
      const guid = p1.props.guid;
      if (guid) {
        const p2 = await Person.findByGuid(guid);
        expect(p1).toEqual(p2);
      } else {
        expect(true).toBeFalsy();
      }
    });

    it("should not get any instance by wrong guid", async () => {
      const p = await Person.findByGuid("asdf");
      expect(p).toBeNull();
    });
  });

  describe("delete", () => {
    it("should return number of deleted persons", async () => {
      await Person.create({ name: "Olaf" });
      const n = await Person.delete({ name: "Olaf" });
      expect(n).toMatchSnapshot();
    });

    it("should return number of deleted olafs", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      const n = await Person.delete({ name: "Olaf" });
      expect(n).toMatchSnapshot();

      let persons = await Person.find();
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });
  });

  describe("find", () => {
    it("should find all persons by name", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      let persons = await Person.find({ name: "Olaf" });
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });

    it("should find all persons", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      let persons = await Person.find();
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });

    it("should find no persons", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      let persons = await Person.find({ name: "Hanns" });
      expect(persons).toMatchSnapshot();
    });

    it("should find all persons starting with 'O'", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Ignatz" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olga" });
      let persons = await Person.find({ name: { $sw: "O" } });
      expect(persons.length).toEqual(2);
      expect(
        persons.map(p => {
          delete p.props.guid;
          return p;
        })
      ).toMatchSnapshot();
    });
  });

  describe("findOne", () => {
    it("should find first person with name", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      let person = await Person.findOne({ name: "Hubert" });
      if (person) {
        delete person.props.guid;
        expect(person).toMatchSnapshot();
      } else {
        expect(1).toEqual(0);
      }
    });

    it("should find no persons", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olaf" });
      let person = await Person.findOne({ name: "Hanns" });
      expect(person).toMatchSnapshot();
    });

    it("should find all persons starting with 'O'", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Ignatz" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Olga" });
      let person = await Person.findOne({ name: { $sw: "Olg" } });
      if (person) {
        delete person.props.guid;
        expect(person).toMatchSnapshot();
      } else {
        expect(1).toEqual(0);
      }
    });
  });

  describe("update", () => {
    it("should update person with name olaf", async () => {
      let p = await Person.create({ name: "Olaf" });
      let persons = await Person.update(p.props, { name: "Hanns" });
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });

    it("should update all persons with name olaf", async () => {
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Olaf" });
      await Person.create({ name: "Hubert" });
      await Person.create({ name: "Ignatz" });
      await Person.create({ name: "Olaf" });
      let p = await Person.update({ name: "Olaf" }, { name: "Hanns" });
      p = p.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(p).toMatchSnapshot();
      let persons = await Person.find();
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });

    it("should update person with name olaf", async () => {
      let p = await Person.create({ name: "Olaf", age: 20 });
      let persons = await Person.update(p.props, { name: "Hanns", age: 22 });
      persons = persons.map(p => {
        delete p.props.guid;
        return p;
      });
      expect(persons).toMatchSnapshot();
    });
  });
});
