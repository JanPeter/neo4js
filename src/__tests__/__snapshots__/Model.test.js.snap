// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model create should create a instance of specific model 1`] = `
ModelInstance {
  "props": Object {
    "age": 21,
    "name": "Paul",
  },
}
`;

exports[`Model delete should return number of deleted olafs 1`] = `3`;

exports[`Model delete should return number of deleted olafs 2`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Hubert",
    },
  },
]
`;

exports[`Model delete should return number of deleted persons 1`] = `1`;

exports[`Model find should find all persons 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hubert",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
]
`;

exports[`Model find should find all persons by name 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
]
`;

exports[`Model find should find all persons starting with 'O' 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Olaf",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Olga",
    },
  },
]
`;

exports[`Model find should find no persons 1`] = `Array []`;

exports[`Model update should update all persons with name olaf 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
]
`;

exports[`Model update should update all persons with name olaf 2`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hubert",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Ignatz",
    },
  },
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
]
`;

exports[`Model update should update person with name olaf 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "name": "Hanns",
    },
  },
]
`;
